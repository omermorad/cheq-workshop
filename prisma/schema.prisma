generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x", "debian-openssl-3.0.x", "linux-musl", "linux-musl-openssl-3.0.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DB_URL")
}

model User {
  id        String    @id @default(uuid())
  name      String
  timezone  String    @default("UTC")
  gender    Gender
  email     String    @unique @db.VarChar(255)
  createdAt DateTime? @default(now()) @map("created_at")

  @@map("users")
}

enum Gender {
  Male
  Female
  Other
}
