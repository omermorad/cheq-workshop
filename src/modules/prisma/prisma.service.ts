import { Injectable, OnModuleInit, OnApplicationShutdown } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit, OnApplicationShutdown {
  public async onModuleInit(): Promise<void> {
    await this.$connect();
  }

  public async onApplicationShutdown(): Promise<void> {
    await this.$disconnect();
  }
}
