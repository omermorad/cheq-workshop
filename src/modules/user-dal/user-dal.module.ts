import { PrismaModule } from '@app/prisma/prisma.module';
import { Module } from '@nestjs/common';
import { UserDal } from './user.dal';

@Module({
  imports: [PrismaModule],
  providers: [UserDal],
  exports: [UserDal],
})
export class UserDalModule {}
