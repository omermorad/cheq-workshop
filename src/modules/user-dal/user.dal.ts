import { Injectable } from '@nestjs/common';
import { User } from '@prisma/client';
import { PrismaService } from '@app/prisma/prisma.service';
import { CreateUserDao } from '@app/app/user-api/types/user.schemas';

@Injectable()
export class UserDal {
  public constructor(private readonly prisma: PrismaService) {}

  public getUserById(id: string): Promise<User | null | never> {
    return this.prisma.user.findUnique({ where: { id } });
  }

  public async createUser(userDao: CreateUserDao): Promise<User | never> {
    return this.prisma.user.create({ data: userDao });
  }

  public async emailExists(email: string): Promise<boolean> {
    const found = await this.prisma.user.findUnique({ where: { email } });
    return Boolean(found);
  }

  public async getAllUsers(): Promise<User[]> {
    return this.prisma.user.findMany();
  }

  public async getUsersByEmailSuffix(domain: string): Promise<{ email: string }[]> {
    return this.prisma.user.findMany({ where: { email: { endsWith: `@${domain}` } }, select: { email: true } });
  }
}
