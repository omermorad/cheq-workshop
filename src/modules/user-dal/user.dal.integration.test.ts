import { Test, TestingModule } from '@nestjs/testing';
import { UserDal } from '@app/user-dal/user.dal';
import { PrismaService } from '@app/prisma/prisma.service';
import { Gender, User } from '@prisma/client';
import { UserDalModule } from '@app/user-dal/user-dal.module';
import { randomEmail } from '../../../test/seeder/functions';

describe('Creating a Random User Integration Test', () => {
  let underTest: UserDal;
  let moduleRef: TestingModule;
  let prisma: PrismaService;

  beforeAll(async () => {
    moduleRef = await Test.createTestingModule({ imports: [UserDalModule] }).compile();
    underTest = moduleRef.get(UserDal);
    prisma = moduleRef.get(PrismaService);
  });

  describe('Email exists', () => {
    const email = randomEmail();

    beforeAll(async () => {
      await prisma.user.create({
        data: { email, name: 'dummy', gender: Gender.Other, timezone: 'Timezone' },
      });
    });

    it('should throw an error because email exists', async () => {
      const promise = underTest.createUser({
        email,
        name: 'dummy',
        gender: Gender.Other,
        timezone: 'Timezone',
      });

      await expect(promise).rejects.toThrow();
    });

    it('should success because email does no exist', async () => {
      const result = await underTest.createUser({
        email: randomEmail(),
        name: 'dummy',
        gender: Gender.Other,
        timezone: 'Timezone',
      });

      expect(result).toMatchSnapshot<User>({
        email: expect.toBeString(),
        name: 'dummy',
        gender: Gender.Other,
        timezone: 'Timezone',
        id: expect.toBeString(),
        createdAt: expect.toBeDate(),
      });
    });

    afterAll(async () => {
      await moduleRef.close();
    });
  });
});
