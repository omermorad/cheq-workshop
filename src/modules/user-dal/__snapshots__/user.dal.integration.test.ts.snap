// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating a Random User Integration Test Email exists should success because email does no exist 1`] = `
{
  "createdAt": toBeDate<>,
  "email": toBeString<>,
  "gender": "Other",
  "id": toBeString<>,
  "name": "dummy",
  "timezone": "Timezone",
}
`;
